persistent/ring.py

```diff
@@ -0,0 +1,228 @@
+
+@implementer(IRing)
+class _DequeRing(object):
+    """
+    A ring backed by the :class:`collections.deque` class. Operations
+    are a mix of constant and linear time.
+
+    It is available on all platforms.
+    """
+
+    __slots__ = ('ring', 'ring_oids')
+
+    def __init__(self):
+
+        self.ring = deque()
+        self.ring_oids = set()
+
+    def __len__(self):
+        return len(self.ring)
+
+    def __contains__(self, pobj):
+        return pobj._p_oid in self.ring_oids
+
+    def add(self, pobj):
+        self.ring.append(pobj)
+        self.ring_oids.add(pobj._p_oid)
+
+    def delete(self, pobj):
+        # Note that we do not use self.ring.remove() because that
+        # uses equality semantics and we don't want to call the persistent
+        # object's __eq__ method (which might wake it up just after we
+        # tried to ghost it)
+        i = 0 # Using a manual numeric counter instead of enumerate() is much faster on PyPy
+        for o in self.ring:
+            if o is pobj:
+                del self.ring[i]
+                self.ring_oids.discard(pobj._p_oid)
+                return 1
+            i += 1
+
+    def move_to_head(self, pobj):
+        self.delete(pobj)
+        self.add(pobj)
+
+    def delete_all(self, indexes_and_values):
+        for ix, value in reversed(indexes_and_values):
+            del self.ring[ix]
+            self.ring_oids.discard(value._p_oid)
+
+    def __iter__(self):
+        return iter(self.ring)
+
+
+try:
+    from cffi import FFI
+except ImportError: # pragma: no cover
+    _CFFIRing = None
+else:
+
+    import os
+    this_dir = os.path.dirname(os.path.abspath(__file__))
+
+    ffi = FFI()
+    with open(os.path.join(this_dir, 'ring.h')) as f:
+        ffi.cdef(f.read())
+
+    ring = ffi.verify("""
+    #include "ring.c"
+    """, include_dirs=[this_dir])
+
+    _OGA = object.__getattribute__
+    _OSA = object.__setattr__
+
+    #pylint: disable=E1101
+    @implementer(IRing)
+    class _CFFIRing(object):
+        """
+        A ring backed by a C implementation. All operations are constant time.
+
+        It is only available on platforms with ``cffi`` installed.
+        """
+
+        __slots__ = ('ring_home', 'ring_to_obj')
+
+        def __init__(self):
+            node = self.ring_home = ffi.new("CPersistentRing*")
+            node.r_next = node
+            node.r_prev = node
+
+            # In order for the CFFI objects to stay alive, we must keep
+            # a strong reference to them, otherwise they get freed. We must
+            # also keep strong references to the objects so they can be deactivated
+            self.ring_to_obj = dict()
+
+        def __len__(self):
+            return len(self.ring_to_obj)
+
+        def __contains__(self, pobj):
+            return getattr(pobj, '_Persistent__ring', self) in self.ring_to_obj
+
+        def add(self, pobj):
+            node = ffi.new("CPersistentRing*")
+            ring.ring_add(self.ring_home, node)
```

Comment:
OK, I see that it is the result of calling `ffi.verify()`.  It should be named something like `_FFI_RING` to signal its global / constant nature.